### PR TITLE
Feat/ping 2176 change component toggle switch

### PR DESCRIPTION
**Pull request by febrimaulana**
### Summary by CodeRabbit

- Refactor: Updated the `normalizeFontSize` and `normalize` functions in the `fonts` utility module for more precise font size calculations.
- Style: Adjusted the styles of the Toggle Switch component and the `BioAndDMSetting` component to improve user interface consistency.
- Test: Modified test cases for the `normalizeFontSize` function to align with the updated logic.
- Chore: Optimized text dimension calculations in the `usePostDetail` hook, enhancing performance and accuracy.

### CODE REVIEW

**Comment by n-permana** [comment]
Image description   CodeRabbit



Currently reviewing new changes in this PR...

<details>
<summary>Commits</summary>
Files that changed from the base of the PR and between d5d5c6cd3965b58be2200e4977bf138a12f3cd21 and 93f32520c8b4d1a7b1587fc3f820b2f40fbe5045 commits.
</details>

<details>
<summary>Files selected (4)</summary>

* __tests__/utils/fonts.test.js (1)
* src/components/ToggleSwitch/index.tsx (5)
* src/screens/ProfileScreen/elements/BioAndDMSetting.tsx (1)
* src/utils/fonts/index.js (2)
</details>


<details>
<summary>Files ignored due to filter (3)</summary>

* __tests__/screens/FeedScreen/__snapshots__/ContentPoll.test.js.snap
* __tests__/screens/GroupInfo/__snapshots__/GroupInfo.test.js.snap
* __tests__/screens/GroupSetting/__snapshots__/GroupSetting.test.js.snap

</details>





---

File: __tests__/utils/fonts.test.js [hunk 1]
```diff
@@ -19,7 +19,7 @@ describe('Util font should correct', () => {
     })
 
    it('normalizeFontSize should run correctly', () => {
-        expect(normalizeFontSize(14)).toEqual(5.447470817120623)
+        expect(normalizeFontSize(14)).toEqual(12.600000000000001)
```
Comment (n-permana) [inline __tests__/utils/fonts.test.js:22]: Image description   CodeRabbit

The test case for `normalizeFontSize` function has been updated to expect a different value. This change is fine as long as the logic of `normalizeFontSize` function has been changed accordingly and the new expected value is correct. However, it's worth noting that comparing floating point numbers directly can lead to precision issues. It would be better to use a function like `toBeCloseTo` for this comparison.
```diff
-        expect(normalizeFontSize(14)).toEqual(12.600000000000001)
+        expect(normalizeFontSize(14)).toBeCloseTo(12.6)
```
This will ensure that the test doesn't fail due to minor differences in precision.

File: src/utils/fonts/index.js [hunk 2]
```diff
@@ -33,12 +34,8 @@ export const normalize = (size) => {
...
 };
 
-export const normalizeFontSize = (fontSize) => {
-  const referenceHeight = 771;
-  const currentScreenHeight = Dimensions.get('screen').height;
-  const normalizedFontSize = (currentScreenHeight / referenceHeight) * fontSize;
-  return normalizedFontSize;
-};
+export const normalizeFontSize = (size, factor = 0.5) =>
+  size + (dimen.normalizeDimen(size) - size) * factor;
```
Comment (n-permana) [inline src/utils/fonts/index.js:38]: Image description   CodeRabbit

The new implementation of `normalizeFontSize` introduces a dependency on the `dimen.normalizeDimen` function. This change could potentially introduce bugs if the `dimen.normalizeDimen` function is not thoroughly tested or if it does not handle edge cases properly. It's recommended to ensure that `dimen.normalizeDimen` has comprehensive test coverage and handles edge cases appropriately.

Additionally, the introduction of the `factor` parameter without any boundary checks can lead to unexpected results if an inappropriate value is passed. Consider adding checks to ensure that `factor` is within a reasonable range (e.g., 0 to 1).

```diff
- export const normalizeFontSize = (size, factor = 0.5) =>
-   size + (dimen.normalizeDimen(size) - size) * factor;
+ export const normalizeFontSize = (size, factor = 0.5) => {
+   if (factor < 0 || factor > 1) {
+     throw new Error('Factor must be between 0 and 1');
+   }
+   return size + (dimen.normalizeDimen(size) - size) * factor;
+ }
```
This will help prevent potential issues caused by incorrect usage of the `factor` parameter.

File: src/components/ToggleSwitch/index.tsx [hunk 3]
```diff
@@ -11,6 +11,7 @@ import {
   TouchableOpacityProps
 } from 'react-native';
 import dimen from '../../utils/dimen';
+import {normalizeFontSize} from '../../utils/fonts';
```
Comment (n-permana) [inline src/components/ToggleSwitch/index.tsx:14]: Image description   CodeRabbit

The import of `normalizeFontSize` from the `fonts` utility module is a good addition. It helps to maintain consistency in font sizes across different screen sizes.
```diff
+ import {normalizeFontSize} from '../../utils/fonts';
```

File: src/components/ToggleSwitch/index.tsx [hunk 4]
```diff
@@ -33,22 +34,30 @@ const styles = StyleSheet.create({
...
     lineHeight: dimen.normalizeDimen(18),
     textAlign: 'center',
+    textAlignVertical: 'center',
     color: '#f4f3f4',
     position: 'absolute',
-    right: dimen.normalizeDimen(2)
+    top: dimen.normalizeDimen(1),
+    right: dimen.normalizeDimen(2),
+    width: dimen.normalizeDimen(22),
+    height: dimen.normalizeDimen(18)
```
Comment (n-permana) [inline src/components/ToggleSwitch/index.tsx:60]: Image description   CodeRabbit

The changes made to the `textOn` and `textOff` styles are significant. The introduction of `textAlignVertical`, `top`, `left`, `right`, `width`, and `height` properties will help in better positioning and alignment of the text within the toggle switch. However, it's important to ensure that these changes don't break the layout on different screen sizes or orientations.
```diff
+ textAlignVertical: 'center',
+ top: dimen.normalizeDimen(1),
+ left: dimen.normalizeDimen(2),
+ width: dimen.normalizeDimen(22),
+ height: dimen.normalizeDimen(18)
```

File: src/components/ToggleSwitch/index.tsx [hunk 5]
```diff
@@ -69,15 +78,15 @@ const styles = StyleSheet.create({
...
-    fontSize: 12,
+    fontSize: normalizeFontSize(12),
     fontWeight: '400'
   },
   labelRight: {
-    marginLeft: 5,
+    marginLeft: dimen.normalizeDimen(5),
     color: '#FFFFFF',
-    fontSize: 12,
+    fontSize: normalizeFontSize(12),
```
Comment (n-permana) [inline src/components/ToggleSwitch/index.tsx:89]: Image description   CodeRabbit

The changes made to the `labelLeft` and `labelRight` styles are appropriate. Using the `dimen.normalizeDimen` function for the `marginRight` and `marginLeft` properties and the `normalizeFontSize` function for the `fontSize` property will ensure that the labels scale properly on different screen sizes.
```diff
+ marginRight: dimen.normalizeDimen(5),
+ fontSize: normalizeFontSize(12),
+ marginLeft: dimen.normalizeDimen(5),
```

File: src/components/ToggleSwitch/index.tsx [hunk 6]
```diff
@@ -169,7 +178,7 @@ const ToggleSwitch: React.FC<ToggleSwitchProps> = ({
       style={[styles.container, containerStyle]}
       activeOpacity={0.9}
       onPress={onPress}>
-      <Text style={[styles.labelLeft, styleLabelLeft]}>{labelLeft}</Text>
+      {labelLeft && <Text style={[styles.labelLeft, styleLabelLeft]}>{labelLeft}</Text>}
```
Comment (n-permana) [inline src/components/ToggleSwitch/index.tsx:181]: Image description   CodeRabbit

The conditional rendering of the `labelLeft` text component based on its existence is a good practice. This prevents unnecessary rendering and potential errors if `labelLeft` is undefined or null.
```diff
+ {labelLeft && <Text style={[styles.labelLeft, styleLabelLeft]}>{labelLeft}</Text>}
```

File: src/components/ToggleSwitch/index.tsx [hunk 7]
```diff
@@ -214,7 +223,7 @@ const ToggleSwitch: React.FC<ToggleSwitchProps> = ({
           ]}
         />
       </Animated.View>
-      <Text style={[styles.labelRight, styleLabelRight]}>{labelRight}</Text>
+      {labelRight && <Text style={[styles.labelRight, styleLabelRight]}>{labelRight}</Text>}
```
Comment (n-permana) [inline src/components/ToggleSwitch/index.tsx:226]: Image description   CodeRabbit

Similarly, the conditional rendering of the `labelRight` text component based on its existence is also a good practice. This prevents unnecessary rendering and potential errors if `labelRight` is undefined or null.
```diff
+ {labelRight && <Text style={[styles.labelRight, styleLabelRight]}>{labelRight}</Text>}
```

**Review by n-permana — COMMENTED** [review]
Image description   CodeRabbit

<details>
<summary>Commits</summary>
Files that changed from the base of the PR and between d5d5c6cd3965b58be2200e4977bf138a12f3cd21 and 93f32520c8b4d1a7b1587fc3f820b2f40fbe5045 commits.
</details>

<details>
<summary>Files selected (4)</summary>

* __tests__/utils/fonts.test.js (1)
* src/components/ToggleSwitch/index.tsx (5)
* src/screens/ProfileScreen/elements/BioAndDMSetting.tsx (1)
* src/utils/fonts/index.js (2)
</details>


<details>
<summary>Files ignored due to filter (3)</summary>

* __tests__/screens/FeedScreen/__snapshots__/ContentPoll.test.js.snap
* __tests__/screens/GroupInfo/__snapshots__/GroupInfo.test.js.snap
* __tests__/screens/GroupSetting/__snapshots__/GroupSetting.test.js.snap

</details>






<details>
<summary>Files skipped from review due to trivial changes (1)</summary>

* src/screens/ProfileScreen/elements/BioAndDMSetting.tsx

</details>

<details>
<summary>Review comments generated (7)</summary>

* Review: 7
* LGTM: 0

</details>

---

<details>
<summary>Tips</summary>

### Chat with Image description  CodeRabbit Bot (`@coderabbitai`)
- Reply on review comments left by this bot to ask follow-up questions. A review comment is a comment on a diff or a file.
- Invite the bot into a review comment chain by tagging `@coderabbitai` in a reply.

### Code suggestions
- The bot may make code suggestions, but please review them carefully before committing since the line number ranges may be misaligned. 
- You can edit the comment made by the bot and manually tweak the suggestion if it is slightly off.

### Pausing incremental reviews
- Add `@coderabbitai: ignore` anywhere in the PR description to pause further reviews from the bot.

</details>

File: __tests__/hooks/usePostDetail.test.js [hunk 8]
```diff
@@ -84,33 +84,33 @@ describe('usePOstDetail should run correctly', () => {
...
 
     expect(result.current.calculationText(longText)).toEqual({
-      containerComment: -794.8,
-      containerHeight: 1164.8,
-      fontSize: 17.066666666666666,
-      lineHeight: 25.6,
+      containerComment: -461.6,
+      containerHeight: 831.6,
+      fontSize: 14.4,
+      lineHeight: 21.6,
```
Comment (n-permana) [inline __tests__/hooks/usePostDetail.test.js:113]: Image description   CodeRabbit

The changes in the `calculationText` function's return values seem to be significant. It appears that the dimensions and font sizes have been adjusted, which could potentially affect the layout and readability of the text in your application. Please ensure these changes are thoroughly tested on different screen sizes and resolutions to avoid any UI issues.
```diff
-      containerComment: 23.333333333333314,
-      containerHeight: 346.6666666666667,
-      fontSize: 25.6,
-      lineHeight: 46.93333333333333,
+      containerComment: 77.5,
+      containerHeight: 292.5,
+      fontSize: 21.6,
+      lineHeight: 39.6,
```
Also, it would be beneficial to add comments explaining why these specific values were chosen and how they were calculated. This will help future developers understand the reasoning behind these numbers and make it easier for them to maintain or modify this code.

File: __tests__/utils/fonts.test.js [hunk 9]
```diff
@@ -1,4 +1,4 @@
-import {normalize, normalizeFontSize, scaleFontSize} from '../../src/utils/fonts'
+import {normalize, normalizeFontSize} from '../../src/utils/fonts'
```
Comment (n-permana) [inline __tests__/utils/fonts.test.js:1]: Image description   CodeRabbit

The import statement has been modified to remove `scaleFontSize`. If this function is no longer used in the project, this change is appropriate. However, if it's still being used elsewhere, ensure that it's imported correctly in those places.
```diff
- import {normalize, normalizeFontSize, scaleFontSize} from '../../src/utils/fonts'
+ import {normalize, normalizeFontSize} from '../../src/utils/fonts'
```

File: __tests__/utils/fonts.test.js [hunk 10]
```diff
@@ -15,16 +15,11 @@ jest.mock('react-native', () => ({
 
 describe('Util font should correct', () => {
     it('normalize should run correctly', () => {
-        expect(normalize(10)).toEqual(8)
+        expect(normalize(10)).toEqual(9)
     })
 
    it('normalizeFontSize should run correctly', () => {
-        expect(normalizeFontSize(14)).toEqual(5.447470817120623)
+        expect(normalizeFontSize(14)).toEqual(12.600000000000001)
```
Comment (n-permana) [inline __tests__/utils/fonts.test.js:22]: Image description   CodeRabbit

The expected values for `normalize` and `normalizeFontSize` functions have changed. This suggests that the underlying logic of these functions has been altered. Ensure that these new values are correct and that they don't break any existing functionality that relies on these utility functions.
```diff
-        expect(normalize(10)).toEqual(8)
-        expect(normalizeFontSize(14)).toEqual(5.447470817120623)
+        expect(normalize(10)).toEqual(9)
+        expect(normalizeFontSize(14)).toEqual(12.600000000000001)
```
Also, consider using a precision or rounding function to avoid floating point precision issues like `12.600000000000001`.

**Review by n-permana — COMMENTED** [review]
Image description   CodeRabbit

<details>
<summary>Commits</summary>
Files that changed from the base of the PR and between 93f32520c8b4d1a7b1587fc3f820b2f40fbe5045 and fec48edf0cd9922067b5b8dcd496a161891acd8c commits.
</details>

<details>
<summary>Files selected (3)</summary>

* __tests__/hooks/usePostDetail.test.js (1)
* __tests__/utils/fonts.test.js (2)
* src/utils/fonts/index.js (2)
</details>


<details>
<summary>Files ignored due to filter (4)</summary>

* __tests__/screens/FeedScreen/__snapshots__/ContentLink.test.js.snap
* __tests__/screens/FeedScreen/__snapshots__/ContentPoll.test.js.snap
* __tests__/screens/GroupInfo/__snapshots__/GroupInfo.test.js.snap
* __tests__/screens/GroupSetting/__snapshots__/GroupSetting.test.js.snap

</details>






<details>
<summary>Files skipped from review due to trivial changes (1)</summary>

* src/utils/fonts/index.js

</details>

<details>
<summary>Review comments generated (3)</summary>

* Review: 3
* LGTM: 0

</details>

---

<details>
<summary>Tips</summary>

### Chat with Image description  CodeRabbit Bot (`@coderabbitai`)
- Reply on review comments left by this bot to ask follow-up questions. A review comment is a comment on a diff or a file.
- Invite the bot into a review comment chain by tagging `@coderabbitai` in a reply.

### Code suggestions
- The bot may make code suggestions, but please review them carefully before committing since the line number ranges may be misaligned. 
- You can edit the comment made by the bot and manually tweak the suggestion if it is slightly off.

### Pausing incremental reviews
- Add `@coderabbitai: ignore` anywhere in the PR description to pause further reviews from the bot.

</details>

**Review by agit26milan — APPROVED** [review]
